src/db.c

```diff
@@ -1727,6 +1835,22 @@ int expireIfNeeded(redisDb *db, robj *key, int flags) {
     return 1;
 }
 
+/* Increases size of the main db to match desired number. In cluster mode resizes all individual dictionaries for slots that
+ * this node owns. */
+void expandDb(const redisDb *db, uint64_t db_size) {
+    if (server.cluster_enabled) {
+        dict *d;
+        dbIterator dbit;
+        dbInitIterator(&dbit, (redisDb *) db);
+        while ((d = dbNextDict(&dbit))) {
+            /* We don't exact number of keys that would fall into each slot, but we can approximate it, assuming even distribution. */
+            dictExpand(d, (db_size / server.cluster->myself->numslots));
```

Comment:
note that since each dict goes to the next power of two, they're all getting a little bit over allocated.
it's likely that this PR has higher dict related memory overheads compared to before.
it may still be worthwhile since on average we could be wasting about 1/2 - 1 pointer pointer per key (e.g. 8 bytes), and we save 16 bytes.

Reply:
Number of free hash slots in dictionaries should be same before and after this change, statistically speaking. Can you explain how you've come up with 1/2 - 1 pointer estimate, cause I don't see this?

Reply:
maybe i'm wrong. let's take a few examples to see.
1. small db DB with 1000 keys, would have used a LUT of 1024 (wasting 24 pointers), will now have 16384 dbs, each of 4 entries (the minimum), so we have 65536 entries (wasting 64,513 pointers).
2. big db with 1m keys, perfectly evenly distributed, would have used 1048576 (wasting 48,576 pointers), would now have 16384 dbs, each of 64 entries  (wasting 48,576 pointers, exactly the same).
3. let's imagine a database that's not evenly distributed, for simplicity let's assume there are only 2 cluster slots.
so if before we had 1m keys in one dict (wasting 64,513 pointers), and let's assume they're split 1/3 + 2/3, so 524288 slots in one, and 1048576 slots in the other (wasting a sum of 572,864 pointers).

so it's clear that we have a problem with small databases, not sure what we wanna do about it if any.

as for the uneven distribution, maybe we wanna ignore it, and maybe on average it won't happen, and as i said in the previous message, maybe it's still worthwhile considering the savings in this PR.

Reply:
> each of 4 entries (the minimum)

An empty dict has no allocated hashtable. It's only the dict struct itself (7 words).

In a small db with 1000 keys in 1000 different slots, the other 15384 dicts are empty.

Reply:
ohh, right, so just the ones with a key will have 4 entries, so this makes it:
1. small db DB with 1000 keys, would have used a LUT of 1024 (wasting 24 pointers), will now have 1024 dbs (the rest will be empty), each of 4 entries (the minimum), so we have 4096 entries (wasting 3,072 pointers).

Reply:
Worst case 1000 non-empty dicts, 3000 wasted pointers. Best case 1 non-empty dict of capacity 1024, wasting 24 pointers.

Dict entry metadata was 2 pointers per key. For 1000 keys we save 2000 pointers.

Reply:
ok, there's also the other case i described above of uneven distribution, which in the past would't have matter, and now each slot has it's own power of two overhead..

Reply:
Yes, but instead of one big power-of-two, there are multiple small power-of-two overheads, but they sum up to roughly the same number (assuming there are no dicts with only 1-2 keys).

|                                        | unstable | this PR |
|----------------------------------------|----------|---------|
| Three slots with 1000 keys in each     | 4096 = 4*1024 | 3*1024 |
| Three slots with 1000, 2000, 3000 keys | 8192 = 8*1024 | 1024+2048+4096 = 7*1024 |

Reply:
Exactly, power of 2 will be much smaller, proportional to the number of dictionaries. Besides small (<1Mb) static overhead for dict structures themselves, there should be no overhead really. And when cluster mode is disabled it should be same as before this change as we allocate only one dict.